game-core/src/main/java/games/strategy/engine/chat/Chat.java

```diff
@@ -155,140 +51,141 @@ public void ping() {}
     this.chatSoundProfile = chatSoundProfile;
-    this.messengers = messengers;
-    statusManager = new StatusManager(messengers);
-    chatChannelName = ChatController.getChatChannelName(chatName);
-    this.chatName = chatName;
-    sentMessages = new SentMessagesHistory();
-
-    /*
-     * The order of events is significant.
-     *
-     * 1. Register our channel listener. Once the channel is registered, we are guaranteed that
-     * when we receive the response from our init message, our channel subscriber has been added,
-     * and will see any messages broadcasted by the server.
-     *
-     * 2. Call the init message on the server remote. Any add or join messages sent from the server
-     * will wait until we receive the init return value.
-     *
-     * 3. When we receive the init message response, initialize our state and release the latch
-     * so all pending messages will get processed. Messages may be ignored if the server
-     * version is incorrect. This all seems a lot more involved than it needs to be.
-     */
-    final IChatController controller = messengers.getRemoteChatController(chatName);
-    messengers.addChatChannelSubscriber(chatChannelSubscriber, chatChannelName);
-    final Tuple<Map<INode, Tag>, Long> init = controller.joinChat();
-    final Map<INode, Tag> chatters = init.getFirst();
-    nodes = new ArrayList<>(chatters.keySet());
-    chatInitVersion = init.getSecond();
-    latch.countDown();
-    assignNodeTags(chatters);
+    sentMessagesHistory = new SentMessagesHistory();
+    chatters = Optional.ofNullable(chatTransmitter.connect()).orElse(new HashMap<>());
     updateConnections();
   }
 
   private void updateConnections() {
-    synchronized (mutexNodes) {
-      if (nodes == null) {
-        return;
-      }
-      final List<INode> playerNames = new ArrayList<>(nodes);
-      Collections.sort(playerNames);
-      for (final IChatListener listener : listeners) {
-        listener.updatePlayerList(playerNames);
-      }
+    final List<ChatParticipant> playerNames =
+        chatters.keySet().stream()
+            .sorted(Comparator.comparing(c -> c.getPlayerName().getValue()))
+            .collect(Collectors.toList());
+
+    for (final IChatListener listener : listeners) {
+      listener.updatePlayerList(playerNames);
     }
   }
 
-  private void addToNotesMap(final INode node, final Tag tag) {
-    if (tag == Tag.NONE) {
+  @Override
+  public void messageReceived(final String message) {
+    final PlayerName from = MessageContext.getSender().getPlayerName();
+    if (isIgnored(from)) {
       return;
     }
-    final LinkedHashSet<String> current = getTagText(tag);
-    notesMap.put(node, current);
+    chatHistory.add(new ChatMessage(message, from));
+    for (final IChatListener listener : listeners) {
+      listener.addMessage(message, from);
+    }
+    // limit the number of messages in our history.
+    while (chatHistory.size() > 1000) {
+      chatHistory.remove(0);
+    }
```

Comment:
I know that this is kept code, but when `chatHistory` is a `CopyOnWriteArrayList`, `remove(0)` will be a super expensive call.
I'd recommend using guavas [`EvictingQueue`](https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/EvictingQueue.java), it should automatically remove the elements, but is a cyclic buffer.
If thread-safety is a requirement, using `Collections#synchronizedCollection` should do the trick.

Reply:
Ok, I just double checked this actually works: It does!

Reply:
Good suggestion: 7e84adcf9